recipes-extended/quark/files/add-POST-and-v4v-support.patch

```diff
@@ -0,0 +1,319 @@
+Author: Nicholas Tsirakis <tsirakisn@ainfosec.com>
+Date:   Tue Feb 12 10:22:24 2019 -0500
+
+    x64 patch
+
+diff --git a/config.mk b/config.mk
+index 7056241..668b0ef 100644
+--- a/config.mk
++++ b/config.mk
+@@ -4,13 +4,13 @@ VERSION = 0
+ # Customize below to fit your system
+ 
+ # paths
+-PREFIX = /usr/local
++PREFIX ?= /usr
+ MANPREFIX = $(PREFIX)/share/man
+ 
+ # flags
+-CPPFLAGS = -DVERSION=\"$(VERSION)\" -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=700 -D_BSD_SOURCE
+-CFLAGS   = -std=c99 -pedantic -Wall -Wextra -Os
+-LDFLAGS  = -s
++CPPFLAGS ?= -DVERSION=\"$(VERSION)\" -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=700 -D_BSD_SOURCE
++CFLAGS   ?= -std=c99 -pedantic -Wall -Wextra -O
++LDFLAGS  ?=
+ 
+ # compiler and linker
+-CC = cc
++CC ?= gcc
+diff --git a/http.c b/http.c
+index c848e81..f53c25c 100644
+--- a/http.c
++++ b/http.c
+@@ -30,6 +30,7 @@ const char *req_field_str[] = {
+ const char *req_method_str[] = {
+ 	[M_GET]  = "GET",
+ 	[M_HEAD] = "HEAD",
++	[M_POST] = "POST"
+ };
+ 
+ const char *status_str[] = {
+@@ -121,14 +122,37 @@ http_get_request(int fd, struct request *r)
+ 		}
+ 	}
+ 
+-	/* remove terminating empty line */
+-	if (hlen < 2) {
++	/* split header and payload (if applicable)
++	 *
++	 *  expected POST format: HEADER,
++	 *                        two newlines,
++	 *                        PAYLOAD,
++	 *                        two newlines
++	 *
++	 *  expected GET format: HEADER,
++	 *                       two newlines
++	 */
++
++	/* find first occurrence of double newlines */
++	p = strstr(h, "\r\n\r\n");
++	if (!p) {
+ 		return http_send_status(fd, S_BAD_REQUEST);
+ 	}
+-	hlen -= 2;
+ 
+-	/* null-terminate the header */
+-	h[hlen] = '\0';
++	/* keep the first newline */
++	p += 2;
++
++	/* find second occurrence - if exists, POST request, else GET */
++	q = strstr(p, "\r\n\r\n");
++	if (q) {
++		*q = '\0';
++		strncpy(r->payload, p, q-p+1);
++	} else {
++		r->payload[0] = '\0';
++	}
++
++	/* null terminate header; strip second newline */
++	*p = '\0';
+ 
+ 	/*
+ 	 * parse request line
+@@ -270,6 +294,8 @@ http_get_request(int fd, struct request *r)
+ 		memmove(r->field[REQ_HOST], p, q - p + 1);
+ 	}
+ 
++	/* wait until our fd is available for the next operation */
++	wait_fd_ready(fd, true, true);
+ 	return 0;
+ }
+ 
+@@ -355,7 +381,7 @@ http_send_response(int fd, struct request *r)
+ 	struct tm tm;
+ 	size_t len, i;
+ 	off_t lower, upper;
+-	int hasport, ipv6host;
++	int hasport, ipv6host, ret;
+ 	static char realtarget[PATH_MAX], tmptarget[PATH_MAX], t[TIMESTAMP_LEN];
+ 	char *p, *q, *mime;
+ 	const char *vhostmatch, *targethost, *err;
+@@ -604,5 +630,10 @@ http_send_response(int fd, struct request *r)
+ 		}
+ 	}
+ 
+-	return resp_file(fd, RELPATH(realtarget), r, &st, mime, lower, upper);
++	/* wait until our fd is available for the next operation */
++	wait_fd_ready(fd, true, true);
++	ret = resp_file(fd, RELPATH(realtarget), r, &st, mime, lower, upper);
++	wait_fd_ready(fd, true, false);
```

Comment:
I'll be honest here, I'm not sure why this fixes the problem either. There were some definite timing issues here that we could only fix by inserting usleep() calls where the wait_fd_ready() calls are now. Since sleeping is a hacky fix, we tried conforming to the socket API a bit more by running the waits through select() calls instead. AFAIK this is the correct way to wait for a socket's readiness, but I may be wrong.